### PR TITLE
Simplify git restore with PackageResolver

### PR DESCRIPTION
**Pull request by yufeih**
Similar to `FileResolver`, `PackageResolver` is used to resolve package dependencies, this is currently a folder or git.

Removed existing features:

- Dependency lock
- bare repo
- git work tree

This enables the ability to restore packages on demand during build, for better local authoring experience.

https://dev.azure.com/ceapex/Engineering/_workitems/edit/121829/

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5405)

### CODE REVIEW

File: docs/specs/git.yml [hunk 1]
```diff
@@ -403,6 +403,7 @@ repos:
         a.md:
 outputs:
   .errors.log: |
+    ["error","committish-not-found","Can't find branch, tag, or commit 'live' for repo *"]
```
Comment (yufeih) [inline docs/specs/git.yml]: This error was raised from `docfx restore`, previous implementation does not raise this error which should be an bug

Comment (live1206) [reply docs/specs/git.yml]: Is this error duplicated with the one below?

Comment (yufeih) [reply docs/specs/git.yml]: Nop, the error from `docfx build` contains file info

File: src/docfx/restore/PackageFetchOptions.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace Microsoft.Docs.Build
+{
+    [Flags]
+    internal enum PackageFetchOptions
+    {
+        /// <summary>
+        /// Default fetch option.
+        /// </summary>
+        None = 0,
```
Comment (live1206) [inline src/docfx/restore/PackageFetchOptions.cs]: how about just renaming this to `DepthOne`, `FullHistory` and `NotThrow`?

File: src/docfx/restore/Restore.cs [hunk 3]
```diff
@@ -81,5 +75,115 @@ private static async Task<bool> RestoreDocset(string docsetPath, string outputPa
...
+
+        private static void RestorePackages(string docsetPath, Config config, string locale, Repository repository)
+        {
+            using (var packageResolver = new PackageResolver(docsetPath, config))
+            {
+                ParallelUtility.ForEach(
+                    GetPackages(config, locale, repository).Distinct(),
+                    item => packageResolver.DownloadPackage(item.package, item.flags),
+                    Progress.Update,
+                    maxDegreeOfParallelism: 8);
```
Comment (live1206) [inline src/docfx/restore/Restore.cs]: should this be related to `Environment.ProcessorCount` instead of a fixed number like `ParallelUtility`?

Comment (yufeih) [reply src/docfx/restore/Restore.cs]: https://github.com/dotnet/docfx/pull/5009

**Pull request merged**